File: news-api/src/main/java/com/news/config/AsyncConfig.java
package com.news.config;
import lombok.extern.slf4j.Slf4j;
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.annotation.AsyncConfigurer;
import org.springframework.scheduling.annotation.EnableAsync;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.concurrent.Executor;

@Configuration
@EnableAsync
@Slf4j
public class AsyncConfig implements AsyncConfigurer {
    @Bean("event-Executor")
    @Override
    public Executor getAsyncExecutor() {
        int coreCount = Runtime.getRuntime().availableProcessors();
        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
        executor.setCorePoolSize(coreCount);  // 기본 스레드 수
        executor.setMaxPoolSize(coreCount*2);  // 최대 스레드 수
        executor.setQueueCapacity(10);  // 큐의 최대 수용량
        executor.setKeepAliveSeconds(60);
        executor.setWaitForTasksToCompleteOnShutdown(true);
        executor.setAwaitTerminationSeconds(60);
        executor.setThreadNamePrefix("event-");
        executor.initialize();
        return executor;
    }

    @Override
    public AsyncUncaughtExceptionHandler getAsyncUncaughtExceptionHandler() {
        return new CustomAsyncExceptionHandler();
    }

    private static class CustomAsyncExceptionHandler implements AsyncUncaughtExceptionHandler {
        @Override
        public void handleUncaughtException(Throwable ex, Method method, Object... params) {
            log.error("Failed to execute {}", ex.getMessage());
            Arrays.asList(params).forEach(param -> log.error("parameter value = {}", param));
        }
    }
}

File: news-api/src/main/kotlin/com/search/config/async/AsyncConfig.kt
package com.search.config.async

import org.slf4j.LoggerFactory
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.scheduling.annotation.AsyncConfigurer
import org.springframework.scheduling.annotation.EnableAsync
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import java.lang.reflect.Method
import java.util.concurrent.Executor

@Configuration
@EnableAsync
class AsyncConfig : AsyncConfigurer{
    private val log = LoggerFactory.getLogger(this::class.java)
    @Bean("event-Executor")
    override fun getAsyncExecutor(): Executor {
        //val coreCount = Runtime.getRuntime().availableProcessors()
        return ThreadPoolTaskExecutor().apply {
            corePoolSize = 2
            maxPoolSize = 5
            queueCapacity = 5
            keepAliveSeconds = 30
            setWaitForTasksToCompleteOnShutdown(true)
            setAwaitTerminationSeconds(10)
            setThreadNamePrefix("event-")
            initialize()
        }
    }

    override fun getAsyncUncaughtExceptionHandler(): AsyncUncaughtExceptionHandler {
        return CustomAsyncExceptionHandler()
    }

    private inner class CustomAsyncExceptionHandler : AsyncUncaughtExceptionHandler{
        override fun handleUncaughtException(ex: Throwable, method: Method, vararg params: Any?) {
            log.error("Failed to execute ${method.name}: ${ex.message}")
            params.forEach {
                param -> log.error("Parameter value = $param")
            }
        }
    }
}